[orc8r][sms] Add sms_ll package

shaddi opened the pull request:
## Summary

sms_ll is a package for handling low-level encoding/decoding of SMS PDUs
(specifically SM-RP and SM-CP protocols). This package is currently focused on
providing support for MT-SMS from orc8r to AGW and handling delivery reports
from UEs.

## Test Plan

- `go test`
- Manual testing with physical UE (Pixel 3) and comparison with Wireshark traces.

shaddi committed:
Add sms_ll package

sms_ll is a package for handling low-level encoding/decoding of SMS PDUs
(specifically SM-RP and SM-CP protocols). This package is currently focused on
providing support for MT-SMS from orc8r to AGW and handling delivery reports
from UEs.

Signed-off-by: Shaddi Hasan <shasan@fb.com>

xjtian commented on lte/cloud/go/sms_ll/sms_cp.go:
```diff
@@ -0,0 +1,174 @@
+}
+
+func SMSCPError(s string) error {
+	return smsCPError{s, errors.New(s)}
```
You can just return `fmt.Errorf("smscp: %s", field)` here instead of creating the new error type

xjtian commented on lte/cloud/go/sms_ll/sms_cp.go:
```diff
@@ -0,0 +1,174 @@
+	Err   error
+}
+
+func SMSCPError(s string) error {
```
`field string`

xjtian commented on lte/cloud/go/sms_ll/sms_cp.go:
```diff
@@ -0,0 +1,174 @@
+func (cpm CPMessage) MarshalBinary() ([]byte, error) {
+	b := make([]byte, 0)
+	b = append(b, cpm.FirstOctet)
+	b = append(b, cpm.MessageType)
```
`b := []byte{cpm.FirstOctet, cpm.MessageType}`

xjtian commented on lte/cloud/go/sms_ll/sms_cp.go:
```diff
@@ -0,0 +1,174 @@
+
+func (cpm CPMessage) TransactionId() byte {
+	fo := cpm.FirstOctet
+	return fo >> 4
```
`return cpm.FirstOctet >> 4`

xjtian commented on lte/cloud/go/sms_ll/sms_cp.go:
```diff
@@ -0,0 +1,174 @@
+func (cpm CPMessage) ProtocolDisc() byte {
+	fo := cpm.FirstOctet
+	fo &= 0xf
+	return fo
```
inline this

xjtian commented on lte/cloud/go/sms_ll/sms_cp.go:
```diff
@@ -0,0 +1,174 @@
+	return b, nil
+}
+
+func (cpm CPMessage) TransactionId() byte {
```
`GetTransactionID`

xjtian commented on lte/cloud/go/sms_ll/sms_cp.go:
```diff
@@ -0,0 +1,174 @@
+	return fo >> 4
+}
+
+func (cpm CPMessage) ProtocolDisc() byte {
```
`GetProtocolDisc`

xjtian commented on lte/cloud/go/sms_ll/sms_cp.go:
```diff
@@ -0,0 +1,174 @@
+	return fo
+}
+
+func (cpm *CPMessage) SetTransactionId(input byte) error {
```
Can we make this struct immutable instead? Maybe we can have different factory functions to instantiate the thing for each permutation of parameters.

Do we have to always specify txid/disc? Or are these optional parameters?

shaddi replied:
I don't think we need to export this, but we do need to always set these parameters, and different instances of the CP messages will have different transaction IDs.

xjtian commented on lte/cloud/go/sms_ll/sms_cp.go:
```diff
@@ -0,0 +1,174 @@
+	}
+
+	cpm.FirstOctet = byte(input[0])
+	cpm.MessageType = byte(input[1])
```
these should be unnecessary casts

xjtian commented on lte/cloud/go/sms_ll/sms_ll.go:
```diff
@@ -0,0 +1,121 @@
+// Outputs:
+//	- Array of byte array representing the set of fully-encoded CP-DATA(RP-DATA(TPDU)) messages generated
+//	- Error	(if any)
+func GenerateSMSDelivers(message string, from_num string, timestamp time.Time, ref_base byte) ([][]byte, error) {
```
`camelCase`

shaddi replied:
This is intended to be part of the public API, so I think it needs to be TitleCase?

xjtian commented on lte/cloud/go/sms_ll/sms_ll.go:
```diff
@@ -0,0 +1,121 @@
+		if e != nil {
+			return nil, e
+		}
+		reference := byte(int(ref_base) + i)
```
let's declare `ref_base` as `uint8` for clarity

xjtian commented on lte/cloud/go/sms_ll/sms_ll.go:
```diff
@@ -0,0 +1,121 @@
+}
+
+// Helper for creating a RP-DATA message. If is_mt is true, make this mobile terminated.
+func CreateRPDataMessage(is_mt bool, reference byte, data []byte) (*RPMessage, error) {
```
it doesn't look like you call this with `is_mt` set to false, should we just remove this param?

shaddi replied:
We should allow this to be set as in theory one can generate RP-DATA messages for both directions, but the right way to do this is to use the MTI rather than a new boolean flag. I'll update.

xjtian commented on lte/cloud/go/sms_ll/sms_rp.go:
```diff
@@ -0,0 +1,367 @@
+
+func SMSRPError(s string) error {
+	return smsRPError{s, errors.New(s)}
+}
```
same as other error type above

xjtian commented on lte/cloud/go/sms_ll/sms_cp.go:
```diff
@@ -0,0 +1,174 @@
+	switch cpm.MessageType {
+	case CP_DATA:
+		cpm.Length = byte(input[2])
+		cpm.RPDU = input[3:]
```
golang slices reference the underlying array, so this could change if the caller does tricky stuff with the param. do a `copy` from the param slice into `RDPU`

xjtian commented on lte/cloud/go/sms_ll/sms_ll.go:
```diff
@@ -0,0 +1,121 @@
+			return nil, e
+		}
+
+		b, e := rpm.MarshalBinary()
```
`marshaledRPM`

xjtian commented on lte/cloud/go/sms_ll/sms_ll.go:
```diff
@@ -0,0 +1,121 @@
+			return nil, e
+		}
+
+		b, e = cpm.MarshalBinary()
```
`marshaledCPM`

xjtian commented on lte/cloud/go/sms_ll/sms_ll.go:
```diff
@@ -0,0 +1,121 @@
+	output := make([][]byte, 0)
+
+	for i := range tpdus {
+		tp, e := tpdus[i].MarshalBinary()
```
`err` instead of `e` to keep with idiomatic Go

xjtian commented on lte/cloud/go/sms_ll/sms_rp.go:
```diff
@@ -0,0 +1,367 @@
+	if len(data) > 232 { // TS24.011 8.2.5.3
+		return nil, SMSRPError("UserData-Element too long (>232 bytes)")
+
+	}
```
validation at the top

xjtian commented on lte/cloud/go/sms_ll/sms_rp.go:
```diff
@@ -0,0 +1,367 @@
+	rpue.Length = byte(len(data))
+	rpue.TPDU = data
+
+	return rpue, nil
```
inline the return:


```
if ... {
  ...
}

return RPUserElement{
  IEI: ...,
  Length: ...,
  TPD: ...,
}, nil
```

xjtian commented on lte/cloud/go/sms_ll/sms_rp.go:
```diff
@@ -0,0 +1,367 @@
+func (rpadde RPAddressElement) MarshalBinary() ([]byte, error) {
+	rpadde_len := 0
+	b := make([]byte, 0, 11)
+	b = append(b, byte(rpadde_len))
```
`b = append(b, 0)`?

shaddi replied:
Oof. Cleaned up.

xjtian commented on lte/cloud/go/sms_ll/sms_rp.go:
```diff
@@ -0,0 +1,367 @@
+	rpadde_len := 0
+	b := make([]byte, 0, 11)
+	b = append(b, byte(rpadde_len))
+	if rpadde.Length != 0x0 {
```
flip the conditional and early return instead `return []byte{0}, nil`

xjtian commented on lte/cloud/go/sms_ll/sms_rp.go:
```diff
@@ -0,0 +1,367 @@
+type RPAddressElement struct {
+	Length     byte // of the number, not the element! actual size is 1 greater
+	NumberInfo byte // octet 3
+	Number     []byte
```
do you need to export these fields? if you hide these, you could create factory functions that do stuff like length validation instead of doing it in `MarshalBinary` for e.g.

xjtian commented on lte/cloud/go/sms_ll/sms_rp.go:
```diff
@@ -0,0 +1,367 @@
+	rpadde.Length = input[0]
+
+	rpadde.NumberInfo = input[1]
+	rpadde.Number = input[2:int(rpadde.Length)]
```
use `copy`. maybe we need some length validation here just in case?

xjtian commented on lte/cloud/go/sms_ll/sms_rp.go:
```diff
@@ -0,0 +1,367 @@
+
+// The RP-Address-Element refers to the number of the SMSC. In our case, we
+// generally don't use an SMSC, so we just set the number to 11.
+func (rpadde *RPAddressElement) SetFakeNumber() error {
```
make this a flat function so the struct can be immutable (well apart from Unmarshal)

ditch the unnecessary error return and bubble it up

xjtian commented on lte/cloud/go/sms_ll/sms_rp.go:
```diff
@@ -0,0 +1,367 @@
+	n := idx + int(rpue.Length)
+	rpue.TPDU = input[idx:n]
+	idx += n
+	return n, nil
```
are you supposed to return idx instead? if not then line above looks redundant

xjtian commented on lte/cloud/go/sms_ll/sms_rp.go:
```diff
@@ -0,0 +1,367 @@
+func (rpce RPCauseElement) MarshalBinary() ([]byte, error) {
+	b := make([]byte, 0, 2)
+	b = append(b, rpce.Length)
+	b = append(b, rpce.Cause)
```
`b = []byte{rpce.Length, rpce.Cause}`

xjtian commented on lte/cloud/go/sms_ll/sms_rp.go:
```diff
@@ -0,0 +1,367 @@
+	return b, nil
+}
+
+func (rpce *RPCauseElement) UnmarshalBinary(input []byte) (int, error) {
```
Do you specifically want to implement `BinaryMarshaler`/`BinaryUnmarshaler`? If not I would ditch the error return in cases like this when it's never non-nil

xjtian commented on lte/cloud/go/sms_ll/sms_rp.go:
```diff
@@ -0,0 +1,367 @@
+func (rpm RPMessage) MarshalBinary() ([]byte, error) {
+	b := make([]byte, 0, 2) // it's at least length 2...
+	b = append(b, rpm.MTI)
+	b = append(b, rpm.Reference)
```
initialize in place with the 2 members

xjtian commented on lte/cloud/go/sms_ll/sms_ll.go:
```diff
@@ -0,0 +1,121 @@
+		da, _ = CreateRPAddressElement(true)
+	} else { // MO-SMS should have empty orig address
+		oa, _ = CreateRPAddressElement(true)
+		da, _ = CreateRPAddressElement(false)
```
you should just return the value type on these instead of the pointer type. ditch the unnecessary error return value from this function, then

```
if isMT {
  rpm.Ori, rpm.Des = Create(...), Create(...)
} else {
  rpm.Ori, rpm.Des = Create(...), Create(...)
}
```

xjtian commented on lte/cloud/go/sms_ll/sms_ll.go:
```diff
@@ -0,0 +1,121 @@
+	rpm.OriginatorAddress = *oa
+	rpm.DestinationAddress = *da
+
+	rpue, e := CreateRPUserElement(data)
```
return the value type

xjtian commented on lte/cloud/go/sms_ll/sms_ll.go:
```diff
@@ -0,0 +1,121 @@
+	}
+	cpm.MessageType = CP_DATA
+	cpm.Length = byte(len(rpdu))
+	cpm.RPDU = rpdu
```
The guts of this function should be replaced by a call to a function that looks something like

```
func CreateCPMessage(txID, disc, messageType, length, rdpu)
```

xjtian commented on lte/cloud/go/sms_ll/sms_ll.go:
```diff
@@ -0,0 +1,121 @@
+		// Create a normal "fake" element
+		rpadde.SetFakeNumber()
+	}
+	return rpadde, nil
```
```
if empty {
  return RPAddressElement{Length: 0}
} else {
  return NewFakeRPElement()
}
```

xjtian reviewed (changes requested):
Mostly little stuff, I'm going to make the reasonable assumption that you know what you're doing r.e. the protocol things.

Not totally through the diff since I'm out of time tonight, but I figure I gave you enough to chip at until I get around to another review :)

Generally:

- Audit what you need to export from the package and the declared types. Do external callers need access to the internal fields and all declared methods of the structs, or do they just need `Marshal/UnmarshalBinary`? How many of these functions need to be called by external clients?
- `err` instead of `e` for errors by convention
- Audit your functions and get rid of unnecessary error returns, unless you're implementing a specific interface that requires it
- `camelCase` or `TitleCase` for exported symbols, never `snake_case`

shaddi commented:
I believe I've addressed all of the points raised. The structs (and their members) are intended to be public, but I've cleaned up some of the helpers to not export everything.

xjtian commented on lte/cloud/go/sms_ll/sms_cp.go:
```diff
@@ -0,0 +1,184 @@
+	cpm.Length = byte(len(rpdu))
+	cpm.RPDU = make([]byte, len(rpdu))
+	copy(cpm.RPDU, rpdu)
+	return *cpm, nil
```
I think it would be clearer if you gathered all the relevant data in local vars and then initialized the entire struct inline. Also, gather all your validation at the top

```
// validate txId
if txID ... {
  return fmt.Errorf()
}

// validate type
switch messageType {
case ...:
  // empty intentionally
default:
  return CPMessage{}, fmt.Errorf(...)
}

firstOctet := blah
// do bitwise stuff to firstOctet
rdpuCopy = make(...)
copy(...)

return CPMessage{
  FirstOctet: firstOctet,
  MessageType: messageType,
  Length: byte(len(rdpuCopy),
  RPDU: rpduCopy,
}, nil
```

xjtian commented on lte/cloud/go/sms_ll/sms_cp.go:
```diff
@@ -0,0 +1,184 @@
+	case CP_ACK:
+		// No additional data, do nothing
+	default:
+		return nil, SMSCPError("Invalid CP MessageType")
```
put the actual message type in here as well for debuggability (fmt.Sprintf)

xjtian commented on lte/cloud/go/sms_ll/sms_cp.go:
```diff
@@ -0,0 +1,184 @@
+		if _, ok := CP_CAUSE_STR[input[2]]; ok {
+			cpm.Cause = input[2]
+		} else {
+			return SMSCPError("Invalid cause")
```
include the invalid value

xjtian commented on lte/cloud/go/sms_ll/sms_cp.go:
```diff
@@ -0,0 +1,184 @@
+	case CP_ACK:
+		// Do nothing -- no more data
+	default:
+		return SMSCPError("Invalid IE type")
```
ditto

xjtian commented on lte/cloud/go/sms_ll/sms_cp.go:
```diff
@@ -0,0 +1,184 @@
+	cpm.FirstOctet |= input
+
+	return nil
+}
```
these 2 mutating methods aren't used anywhere else other than the factory function so you should explode them into the caller

xjtian commented on lte/cloud/go/sms_ll/sms_cp.go:
```diff
@@ -0,0 +1,184 @@
+}
+
+func (cpm *CPMessage) setProtocolDisc(input byte) error {
+	if input != CP_PROTOCOL_DISC {
```
you control the only callsite so this validation seems overkill

xjtian commented on lte/cloud/go/sms_ll/sms_ll.go:
```diff
@@ -0,0 +1,101 @@
+// Outputs:
+//	- Array of byte array representing the set of fully-encoded CP-DATA(RP-DATA(TPDU)) messages generated
+//	- Error	(if any)
+func GenerateSMSDelivers(message string, from_num string, timestamp time.Time, ref_base uint8) ([][]byte, error) {
```
`camelCase`

xjtian commented on lte/cloud/go/sms_ll/sms_ll.go:
```diff
@@ -0,0 +1,101 @@
+	return tpdus
+}
+
+// Helper for creating a RP-DATA message. If is_mt is true, make this mobile terminated.
```
police the docstring, is_mt is no longer a param

xjtian commented on lte/cloud/go/sms_ll/sms_ll.go:
```diff
@@ -0,0 +1,101 @@
+	if err != nil {
+		return nil, err
+	}
+	return rpm, nil
```
for readability, initialize as much of the struct as possible with the declaration and try to pull all the error-returning blocks near the top

```
msgType, err := ...
if ... {} else if .. {}

ud, err := createRPU...()
if ... {}

rpm := RPMessage {
  MTI: mti,
  Reference: reference,
  UserData: ud,
}
if Direction {} else {}
return rpm, nil
```

xjtian commented on lte/cloud/go/sms_ll/sms_ll.go:
```diff
@@ -0,0 +1,101 @@
+	return output, nil
+}
+
+func createTPDUs(message string, from_num string, timestamp time.Time) []tpdu.TPDU {
```
camelCase

xjtian commented on lte/cloud/go/sms_ll/sms_ll.go:
```diff
@@ -0,0 +1,101 @@
+// Outputs:
+//	- Array of byte array representing the set of fully-encoded CP-DATA(RP-DATA(TPDU)) messages generated
+//	- Error	(if any)
+func GenerateSMSDelivers(message string, from_num string, timestamp time.Time, ref_base uint8) ([][]byte, error) {
```
Do we need to export anything other than this function from the package? Everything else seems like implementation details that a caller otuside this package should not see

shaddi replied:
My thinking is we need to expose at least Unmarshal for all types so callers can decode specific headers, and I'm not sure why we wouldn't want to expose the types and their fields given there's useful information in them (e.g., delivery status, the encapsulated TPDU, etc.)

shaddi replied:
Capturing our offline discussion: we're going to expose only a very simplified public API for this module, so these are all going to be non-exported now.

xjtian commented on lte/cloud/go/sms_ll/sms_rp.go:
```diff
@@ -0,0 +1,365 @@
+	}
+
+	if int(rpadde.Length) - 1 != len(rpadde.Number) {
+		return nil, SMSRPError("Address size mismatch")
```
this validation should be done in `UnmarshalBinary`. You should be able to trust these structs, particularly if you don't export them. ultimately MarshalBinary should not return an error

xjtian commented on lte/cloud/go/sms_ll/sms_rp.go:
```diff
@@ -0,0 +1,365 @@
+			rpadde.Length = input[0]
+			return 1, nil
+		}
+	} else if len(input) < 2 {
```
if len(input) != 1 and len(input) < 2, doesn't this conditional really mean `else if len(input) == 0`? that seems clearer

xjtian commented on lte/cloud/go/sms_ll/sms_rp.go:
```diff
@@ -0,0 +1,365 @@
+
+}
+
+func (rpue RPUserElement) MarshalBinary(msgType byte) ([]byte, error) {
```
remove the unused error return type

xjtian commented on lte/cloud/go/sms_ll/sms_rp.go:
```diff
@@ -0,0 +1,365 @@
+	return b, nil
+}
+
+func (rpue *RPUserElement) UnmarshalBinary(msgType byte, input []byte) (int, error) {
```
error return type is unused

xjtian commented on lte/cloud/go/sms_ll/sms_rp.go:
```diff
@@ -0,0 +1,365 @@
+		}
+	case RP_ERROR:
+		// Do nothing
+		rpm.Cause = *new(RPCauseElement)
```
ditto

xjtian commented on lte/cloud/go/sms_ll/sms_rp.go:
```diff
@@ -0,0 +1,365 @@
+	return RP_MT
+}
+
+func (rpm RPMessage) MsgType() (byte, error) {
```
IMO you should save this into the RPMesage struct even though it's a derived value. Then you can trust your struct when you're serializing it. Do the validation during the unmarshal

shaddi replied:
Unless you feel strongly about this, I prefer derived values (that are more of a convenience than anything) to live outside the struct. Validation is still done in `unserialize`, so the struct is valid, and we can just use these when/if they're needed.

xjtian commented on lte/cloud/go/sms_ll/sms_rp.go:
```diff
@@ -0,0 +1,365 @@
+		// The next two IEs should be adddresses in this case. So, get the lengths and pass to unmarshal
+		rpm.OriginatorAddress = *new(RPAddressElement)
+		rpm.DestinationAddress = *new(RPAddressElement)
+		rpm.UserData = *new(RPUserElement)
```
you don't actually have to initialize these fields. in Go the zero type for a struct is the struct with all its fields set to their zero values: https://medium.com/rungo/structures-in-go-76377cc106a2

xjtian commented on lte/cloud/go/sms_ll/sms_rp.go:
```diff
@@ -0,0 +1,365 @@
+		da_data, err := rpm.DestinationAddress.MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
```
if you move that validation thing to unmarshal, then you can get rid of the err return value and just do `append(b, foo.MarshalBinary()...)`

xjtian commented on lte/cloud/go/sms_ll/sms_rp.go:
```diff
@@ -0,0 +1,365 @@
+		b = append(b, ud_data...)
+	case RP_ERROR:
+		cause_data := rpm.Cause.MarshalBinary()
+		b = append(b, cause_data...)
```
inline the line above into here

xjtian commented on lte/cloud/go/sms_ll/sms_rp.go:
```diff
@@ -0,0 +1,365 @@
+	case RP_ACK:
+		// Do nothing
+	default:
+		return nil, SMSRPError("Invalid RP Message Type")
```
this should be unreachable because of UnmarshalBinary; use if-else instead of switch to bypass the compiler, or panic in the default

xjtian commented on lte/cloud/go/sms_ll/sms_rp.go:
```diff
@@ -0,0 +1,365 @@
+func (rpm RPMessage) MarshalBinary() ([]byte, error) {
+	b := []byte{rpm.MTI, rpm.Reference}
+
+	rpmt, err := rpm.MsgType()
```
if you're calling MarshalBinary() on one of these structs, there should never be an error returned from this method. you can make this explicit with `rpmt, _ := ...`, or you can panic on err. MarshalBinary() should not return an error

xjtian reviewed (changes requested):
From POV of the SMS cloud service that would use this package: only `GenerateSMSDelivers` needs to be exposed (exported) to me, everything else should be hidden.

I think the flow suffers from scattered validation. I would expect that a struct I get after a successful `UnmarshalBinary` would always give me a succesful `MarshalBinary`, because these are my structs and I should be able to trust my validation code. IMO any `MarshalBinary` methods that return ([]byte, error) instead of just []byte are smelly.

Be defensive along the path of initialization but don't bleed that anywhere else. These should all be internal structs that you have full control over.
